lib/ValidationError.ts

```diff
@@ -0,0 +1,65 @@
+import * as zod from 'zod';
+
+import { joinPath } from './utils/joinPath';
+
+const PREFIX_COPY = 'Validation error';
+const MAX_ISSUES_IN_REASON = 10;
+
+export class ValidationError extends Error {
+  details: Array<Zod.ZodIssue>;
+
+  constructor(
+    message: string,
+    options: {
+      details: Array<Zod.ZodIssue>;
+    }
+  ) {
+    super(message);
+    this.details = options.details;
+  }
+}
+
+/**
+ * Converts the supplied ZodError to ValidationError.
+ * @param zodError {zod.ZodError}
+ * @return {ValidationError}
+ */
+export function fromZodError(zodError: zod.ZodError): ValidationError {
+  const reason = zodError.errors
+    // limit max number of issues printed in the reason section
+    .slice(0, MAX_ISSUES_IN_REASON)
+    // format error message
+    .map((issue) => {
+      const { message, path } = issue;
+
+      if (path.length > 0) {
+        return message + ' at "' + joinPath(path) + '"';
+      }
+
+      return message;
+    })
+    // concat as string
+    .join('; ');
+
+  const message = reason ? [PREFIX_COPY, reason].join(': ') : PREFIX_COPY;
```

Comment:
[suggestion]: I would leave this functionality to the consumer apps. A simple `Validation Error(s)` on the `error.message` field together with the extra `[{message: string; path: string[]}]` array on the  `error.details` should be enough for consumers to build their error messages according to their requirements. I also think the number of errors should not be sliced at this point.

Reply:
Given that one does not contradict the other, i.e. having a ready-made error message does not prevent you from compiling your own message, I would prefer to leave it as is.

In addition I am now accepting formatting options to further configure the error message.